geometry_constructor/validators.py

```diff
@@ -0,0 +1,119 @@
+"""Input validators to be used on QML input fields"""
+from geometry_constructor.instrument_model import InstrumentModel
+from PySide2.QtCore import Property, Signal
+from PySide2.QtGui import QValidator
+
+
+class ValidatorOnInstrumentModel(QValidator):
+    """
+    Base class for validators that check an indexed item against other components in an InstrumentModel
+
+    Exposes properties that can be set in QML for the component index, and instrument model.
+    In QML these can be set with 'myindex' and 'model'
+    A signal 'validationFailed' is included so that error messages can be displayed in the UI to explain the failure.
+
+    Getters and setters, while un-pythonic, are required for the Qt properties system
+    https://wiki.qt.io/Qt_for_Python_UsingQtProperties
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.instrument_model = InstrumentModel()
+        self.model_index = -1
+
+    def get_index(self):
+        return self.model_index
+
+    def set_index(self, val: int):
+        self.model_index = val
+
+    @Signal
+    def index_changed(self):
+        pass
+
+    myindex = Property(int, get_index, set_index, notify=index_changed)
+
+    def get_model(self):
+        return self.instrument_model
+
+    def set_model(self, val: InstrumentModel):
+        self.instrument_model = val
+
+    @Signal
+    def model_changed(self):
+        pass
+
+    model = Property('QVariant', get_model, set_model, notify=model_changed)
+
+    validationFailed = Signal()
+
+
+class TransformParentValidator(ValidatorOnInstrumentModel):
+    """
+    Validator to prevent circular transform parent dependencies being created in an instrument model
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def validate(self, input: str, pos: int):
+        """
+        Validates the input as the name of a component's transform parent to check it won't cause a circular dependency
+
+        The signal 'validationFailed' is emitted only if a circular dependency is found
+        (http://doc.qt.io/qt-5/qvalidator.html#validate)
+        """
+        parents = {}
+
+        candidate_parent_index = -1
+        for component in self.instrument_model.components:
```

Comment:
Could you make this function easier to follow? What is the role of each of these `for` loops, could they be extracted to separate functions with elucidating names?